RFCs/0009-Fn::Map.md

```diff
@@ -0,0 +1,898 @@
+    * **This field is only required when Fn::Map should generate an object of key value pairs, for example in the Resources or Outputs section of a template. It should not be specified when using Fn::Map to generate an array, for example a list type attribute for a Resource Property.** 
+    * The value **should be interpolated with an index** using `Fn::Sub` intrinsic function.
+
+
+## Fn::Merge
+Fn::Merge takes an array of objects as input and merges them into the parent object
+```json
+{
+  "Fn::Merge": [
+    {
+      "key1": { },
+      "key2": { }
+    },
+    {
+      "key3": { }
+    }
+  ],
+  "key4": { }
+}
+```
+
+Result:
+```json
+{
+  "key1": { },
+  "key2": { },
+  "key3": { },
+  "key4": { }
+}
+```
+
+Fn::Merge will throw an error when there is a key collision. The below examples will cause an error:
+
+```json
+{
+  "Fn::Merge": [
+    {
+      "key1": {}
+    },
+    {
+      "key1": { }
+    }
+  ]
+}
+```
+```json
+{
+  "Fn::Merge": [
+    {
+      "key1": {}
+    }
+  ],
+  "key1": { }
+}
+```
+
+# Examples
+### Usage in Resources Section
+#### Use case: Replicate a single Resource
+* Note that the **`Key`** parameter is provided.
+
+In Json:
+```json
+{
+  "AWSTemplateFormatVersion": "2010-09-09",
+  "Description": "EC2 Instances with different AMIs",
+  "Resources": {
+    "Fn::Merge": [
+      {
+        "Fn::Map": {
+          "Index": "i",
+          "Value": "x",
+          "Collection": ["ami-1", "ami-2", "ami-3"],
+          "Fragment": {
+            "Type": "AWS::EC2::Instance",
+            "Properties": {
+              "InstanceType": "m1.small",
+              "ImageId": {"Ref": "x"}
+            }
+          },
+          "Key": {
+            "Fn::Sub": "Instance${i}"
+          }
+        }
+      }
+    ],
+    "MyS3Bucket": {
+      "Type": "AWS::S3::Bucket"
+    },
+    "MyQueue": {
+      "Type": "AWS::SQS::Queue"
+    }
+  }
+}
+```
+* In Yaml
+```yaml
+AWSTemplateFormatVersion: "2010-09-09"
+Description: EC2 Instances with different AMIs
+Resources:
+  Fn::Merge:
+    - Fn::Map:
+        Index: i
+        Value: x
+        Collection: [ "ami-1", "ami-2", "ami-3" ]
+        Fragment:
+          Type: AWS::EC2::Instance
+          Properties:
+            InstanceType: m1.small
+            ImageId: !Ref x
+        Key:
+          Fn::Sub: Instance${i}
+  MyS3Bucket:
+    Type: AWS::S3::Bucket
+  MyQueue:
+    Type: AWS::SQS::Queue
+
+```
+* The above templates would be equivalent to the one below:
+```yaml
+AWSTemplateFormatVersion: '2010-09-09'
+Description: "EC2 Instances with different AMIs"
+Resources:
+  Instance0:
+    Type: AWS::EC2::Instance
+    Properties:
+      InstanceType: "m1.small"
+      ImageId: "ami-1"
+  Instance1:
+    Type: AWS::EC2::Instance
+    Properties:
+      InstanceType: "m1.small"
+      ImageId: "ami-2"
+  Instance2:
+    Type: AWS::EC2::Instance
+    Properties:
+      InstanceType: "m1.small"
+      ImageId: "ami-3"
+```
```

Comment:
The above examples also contain `MyS3Bucket` and `MyQueue`, so I think we should include those here too (or remove them from the previous).

Reply:
You're right!